Branch-SDK/src/io/branch/referral/RemoteInterface.java

```diff
@@ -26,128 +27,183 @@
+            boolean isRetryableStatusCode = statusCode >= 500;
+
+            // Retry request before processing
+            if ( (retryNumber < prefHelper_.getRetryCount()) && isRetryableStatusCode) {
+                retryNumber +=1;
+                make_restful_get(url, tag, timeout, retryNumber, log);
+            } else {
+                return processEntityForJSON(response.getEntity(),
+                        response.getStatusLine().getStatusCode(), tag, log, null);
+            }
+
+        } catch (ClientProtocolException ex) {
+            if (log) {
+                PrefHelper.Debug(getClass().getSimpleName(), "Client protocol exception: " + ex.getMessage());
+            }
+        } catch (SocketException ex) {
+            if (log) {
+                PrefHelper.Debug(getClass().getSimpleName(), "Http connect exception: " + ex.getMessage());
+            }
+            return new ServerResponse(tag, NO_CONNECTIVITY_STATUS);
+        } catch (UnknownHostException ex) {
+            if (log) {
+                PrefHelper.Debug(getClass().getSimpleName(), "Http connect exception: " + ex.getMessage());
+            }
+            return new ServerResponse(tag, NO_CONNECTIVITY_STATUS);
+        } catch (IOException ex) {
+            if (log) {
+                PrefHelper.Debug(getClass().getSimpleName(), "IO exception: " + ex.getMessage());
+            }
+            return new ServerResponse(tag, 500);
+        }
+        return null;
+    }
+    //endregion
+
+    //region POST requests
+    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout) {
+        return make_restful_post(body, url, tag, timeout, 0, true, null);
+    }
+
+    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, BranchLinkData linkData) {
+        return make_restful_post(body, url, tag, timeout, 0, true, linkData);
+    }
+
+    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, boolean log) {
+        return make_restful_post(body, url, tag, timeout, 0, log, null);
+    }
+
+    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout,
+                                            int retryNumber, boolean log, BranchLinkData linkData) {
+        try {
+            if (body.has("app_id") && body.getString("app_id").equals(PrefHelper.NO_STRING_VALUE)) {
+                return new ServerResponse(tag, NO_API_KEY_STATUS);
+            }
+
+            body.put("sdk", "android" + SDK_VERSION);
+            if (log) {
+                PrefHelper.Debug(BranchConfig.TAG, "posting to " + url);
+                PrefHelper.Debug(BranchConfig.TAG, "Post value = " + body.toString(4));
+            }
+            HttpPost request = new HttpPost(url);
+            request.setEntity(new ByteArrayEntity(body.toString().getBytes("UTF8")));
+            request.setHeader("Content-type", "application/json");
```

Comment:
Content-Type should be PascalCased, though according to the RFC, the server should accept it either way.